File: src/ecs/entity.cpp
// Copyright 2025 Valentin Namleev

#include "feyerverx/ecs/entity.hpp"
#include "feyerverx/ecs/registry.hpp"

namespace feyerverx::ecs {
Entity::Entity() : m_ID(NullEntityID) {}
Entity::Entity(EntityID _ID, Registry *const _Registry) : m_ID(_ID), m_Registry(_Registry) {}
EntityID Entity::ID() const { return m_ID; }
} // namespace feyerverx::ecs

File: src/ecs/entity.zig
pub const Registry = @import("registry.zig").Registry;

pub const EntityID = u16;
pub const Entity = struct {
    entity_id: EntityID,
    registry: *Registry,

    pub fn create(entity_id: EntityID, registry: *Registry) Entity {
        return Entity{
            .entity_id = entity_id,
            .registry = registry,
        };
    }

    pub fn add(self: Entity, comptime T: type, value: T) !void {
        try self.registry.add(self.entity_id, T, value);
    }

    pub fn get(self: Entity, comptime T: type) ?*T {
        return self.registry.get(self.entity_id, T);
    }
};
